File: java/core/libjoynr/src/main/java/io/joynr/proxy/DiscoveryResult.java
/*
 * #%L
 * %%
 * Copyright (C) 2019 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package io.joynr.proxy;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Optional;

import io.joynr.arbitration.ArbitrationConstants;
import joynr.types.CustomParameter;
import joynr.types.DiscoveryEntry;
import joynr.types.DiscoveryEntryWithMetaInfo;
import joynr.types.Version;

/**
 * This class represents a List of DiscoveryEntries resulting from an arbitration.
 * It provides several filtering methods that allow to retrieve DiscoveryEntries by
 * different criteria.
 */
public class DiscoveryResult {

    private List<DiscoveryEntry> discoveryEntries;

    /**
     * Constructor for internal use only.
     *
     * @param discoveryEntries DiscoveryEntries for initialization.
     */
    public DiscoveryResult(Collection<DiscoveryEntryWithMetaInfo> discoveryEntries) {
        this.discoveryEntries = new ArrayList<>();
        for (DiscoveryEntry entry : discoveryEntries) {
            this.discoveryEntries.add(new DiscoveryEntry(entry));
        }
    }

    /**
     * Returns the DiscoveryEntry with the most recent last seen date from this DiscoveryResult.
     * <p>
     * The last seen date of a provider is periodically updated by joynr at a configurable interval, property:
     * {@link io.joynr.runtime.SystemServicesSettings#PROPERTY_CAPABILITIES_FRESHNESS_UPDATE_INTERVAL_MS}.
     * <p>
     * Providers with a very outdated last seen date are most likely not reachable anymore, for example because of
     * connectivity issues or a crash of the provider application.
     *
     * @return Returns the DiscoveryEntry with the most recent last seen date from this DiscoveryResult.
     */
    public DiscoveryEntry getLastSeen() {
        DiscoveryEntry lastSeenEntry = discoveryEntries.get(0);
        for (DiscoveryEntry entry : discoveryEntries) {
            if (entry.getLastSeenDateMs() > lastSeenEntry.getLastSeenDateMs()) {
                lastSeenEntry = entry;
            }
        }
        return lastSeenEntry;
    }

    /**
     * Returns the DiscoveryEntry with the highest priority from this DiscoveryResult.
     * <p>
     * The priority of a provider can be set by an application when registering the provider, see
     * {@link joynr.types.ProviderQos#setPriority(Long)}.
     *
     * @return Returns the DiscoveryEntry with the highest priority from this DiscoveryResult.
     * @see joynr.types.ProviderQos#setPriority(Long)
     */
    public DiscoveryEntry getHighestPriority() {
        DiscoveryEntry entryWithHighestPriority = discoveryEntries.get(0);
        for (DiscoveryEntry entry : discoveryEntries) {
            if (entry.getQos().getPriority() > entryWithHighestPriority.getQos().getPriority()) {
                entryWithHighestPriority = entry;
            }
        }
        return entryWithHighestPriority;
    }

    /**
     * Returns the DiscoveryEntry with the highest Version number from this DiscoveryResult.
     *
     * @return Returns the DiscoveryEntry with the highest Version from this DiscoveryResult.
     */
    public DiscoveryEntry getLatestVersion() {
        DiscoveryEntry entryWithLatestVersion = discoveryEntries.get(0);
        for (DiscoveryEntry entry : discoveryEntries) {
            Version currentLatest = entryWithLatestVersion.getProviderVersion();
            Version inspectedVersion = entry.getProviderVersion();
            if (inspectedVersion.getMajorVersion() > currentLatest.getMajorVersion()) {
                entryWithLatestVersion = entry;
            } else if (inspectedVersion.getMajorVersion().equals(currentLatest.getMajorVersion())) {
                if (inspectedVersion.getMinorVersion() > currentLatest.getMinorVersion()) {
                    entryWithLatestVersion = entry;
                }
            }
        }
        return entryWithLatestVersion;
    }

    /**
     * Returns a Collection with all DiscoveryEntries in this DiscoveryResult.
     *
     * @return Returns a Collection with all DiscoveryEntries in this DiscoveryResult.
     */
    public Collection<DiscoveryEntry> getAllDiscoveryEntries() {
        return discoveryEntries;
    }

    /**
     * Returns the DiscoveryEntry with the given participantId, or null if it doesn't exist.
     *
     * @param participantId ParticipantId of the required DiscoveryEntry.
     * @return Returns the DiscoveryEntry with the given participantId, or null if it doesn't exist.
     */
    public DiscoveryEntry getParticipantId(String participantId) {
        for (DiscoveryEntry entry : discoveryEntries) {
            if (entry.getParticipantId().equals(participantId)) {
                return entry;
            }
        }
        return null; //Or throw exception?
    }

    /**
     * Returns a Collection of DiscoveryEntries that have the given keyword in their custom parameters.
     * <p>
     * A keyword for a provider can be set by an application when registering the provider by setting a custom parameter
     * in the {@link joynr.types.ProviderQos} with the special key {@link ArbitrationConstants#KEYWORD_PARAMETER}, see
     * {@link joynr.types.ProviderQos#setCustomParameters(CustomParameter[])}.
     *
     * @param keyword The custom parameters of the required DiscoveryEntries should contain this keyword.
     * @return Returns a Collection of DiscoveryEntries that have the given keyword in their custom parameters.
     * @see joynr.types.ProviderQos#setCustomParameters(CustomParameter[])
     */
    public Collection<DiscoveryEntry> getWithKeyword(String keyword) {
        List<DiscoveryEntry> entriesWithKeyword = new ArrayList<>();
        for (DiscoveryEntry entry : discoveryEntries) {
            Optional<CustomParameter> keywordParameter = findQosParameter(entry,
                                                                          ArbitrationConstants.KEYWORD_PARAMETER);
            if (keywordParameter.isPresent() && keywordParameter.get().getValue().equals(keyword)) {
                entriesWithKeyword.add(entry);
            }
        }
        return entriesWithKeyword;
    }

    private Optional<CustomParameter> findQosParameter(DiscoveryEntry discoveryEntry, String parameterName) {
        for (CustomParameter parameter : discoveryEntry.getQos().getCustomParameters()) {
            if (parameterName.equals(parameter.getName())) {
                return Optional.of(parameter);
            }
        }
        return Optional.empty();

    }
}

File: cpp/libjoynr/joynr-messaging/DiscoveryResult.cpp
/*
 * #%L
 * %%
 * Copyright (C) 2020 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
#include "joynr/DiscoveryResult.h"

#include "joynr/DiscoveryQos.h"
#include "joynr/types/Version.h"

namespace joynr
{

DiscoveryResult::DiscoveryResult(const std::vector<types::DiscoveryEntry>& discoveryEntries)
        : _discoveryEntries(discoveryEntries)
{
}

const std::vector<joynr::types::DiscoveryEntry>& DiscoveryResult::getAllDiscoveryEntries() const
{
    return _discoveryEntries;
}

boost::optional<types::DiscoveryEntry> DiscoveryResult::getLastSeen() const
{
    boost::optional<joynr::types::DiscoveryEntry> entryWithLastSeenDate;
    if (_discoveryEntries.empty()) {
        return entryWithLastSeenDate;
    }

    entryWithLastSeenDate = _discoveryEntries.front();
    for (const auto& entry : _discoveryEntries) {
        if (entry.getLastSeenDateMs() > entryWithLastSeenDate.value().getLastSeenDateMs()) {
            entryWithLastSeenDate = entry;
        }
    }
    return entryWithLastSeenDate;
}

boost::optional<types::DiscoveryEntry> DiscoveryResult::getWithHighestPriority() const
{
    boost::optional<joynr::types::DiscoveryEntry> entryWithHighestPriority;
    if (_discoveryEntries.empty()) {
        return entryWithHighestPriority;
    }

    entryWithHighestPriority = _discoveryEntries.front();
    for (const auto& entry : _discoveryEntries) {
        if (entry.getQos().getPriority() >
            entryWithHighestPriority.value().getQos().getPriority()) {
            entryWithHighestPriority = entry;
        }
    }
    return entryWithHighestPriority;
}

boost::optional<types::DiscoveryEntry> DiscoveryResult::getWithLatestVersion() const
{
    boost::optional<joynr::types::DiscoveryEntry> entryWithLatestVersion;
    if (_discoveryEntries.empty()) {
        return entryWithLatestVersion;
    }

    entryWithLatestVersion = _discoveryEntries.front();
    for (const auto& entry : _discoveryEntries) {
        joynr::types::Version currentVersion = entryWithLatestVersion.value().getProviderVersion();
        joynr::types::Version inspectedVersion = entry.getProviderVersion();

        if (inspectedVersion.getMajorVersion() > currentVersion.getMajorVersion()) {
            entryWithLatestVersion = entry;
        } else if (inspectedVersion.getMajorVersion() == currentVersion.getMajorVersion()) {
            if (inspectedVersion.getMinorVersion() > currentVersion.getMinorVersion()) {
                entryWithLatestVersion = entry;
            }
        }
    }
    return entryWithLatestVersion;
}

boost::optional<types::DiscoveryEntry> DiscoveryResult::getWithParticipantId(
        const std::string& participantId) const
{
    boost::optional<joynr::types::DiscoveryEntry> entryWithParticipantId;
    if (_discoveryEntries.empty()) {
        return entryWithParticipantId;
    }

    for (const auto& entry : _discoveryEntries) {
        if (entry.getParticipantId() == participantId) {
            entryWithParticipantId = entry;
            break;
        }
    }
    return entryWithParticipantId;
}

std::vector<joynr::types::DiscoveryEntry> DiscoveryResult::getWithKeyword(
        const std::string& keyword) const
{
    std::vector<joynr::types::DiscoveryEntry> entriesWithKeyword;
    for (const auto& entry : _discoveryEntries) {
        boost::optional<joynr::types::CustomParameter> keywordParameter =
                findQosParameter(entry, DiscoveryQos::KEYWORD_PARAMETER());
        if (keywordParameter && keywordParameter.get().getValue() == keyword) {
            entriesWithKeyword.push_back(entry);
        }
    }
    return entriesWithKeyword;
}

boost::optional<joynr::types::CustomParameter> DiscoveryResult::findQosParameter(
        const types::DiscoveryEntry& discoveryEntry,
        const std::string& parameterName) const
{
    boost::optional<joynr::types::CustomParameter> customParameter;
    for (const auto& parameter : discoveryEntry.getQos().getCustomParameters()) {
        if (parameterName == parameter.getName()) {
            customParameter = parameter;
            break;
        }
    }
    return customParameter;
}

} // namespace joynr
